File: src/calculus_optimization_easy.py
"""
calculus - optimization (easy)
Generated: 2026-02-22T05:08:37.011165
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 4)
    
    if problem_type == 1:
        # Type 1: Find critical points of a simple quadratic (1000-1200)
        a_val = choice([-2, -1, 1, 2])
        h = nonzero(-5, 5)
        k = randint(-10, 10)
        # f(x) = a(x - h)^2 + k, critical point at x = h
        expr = a_val * (x - h)**2 + k
        expr_expanded = expand(expr)
        
        derivative = diff(expr_expanded, x)
        critical_pt = h
        
        return problem(
            question=f"Find the critical point of $f(x) = {latex(expr_expanded)}$",
            answer=critical_pt,
            difficulty=(1000, 1200),
            topic="calculus/optimization",
            solution=steps(
                f"Find the derivative: $f'(x) = {latex(derivative)}$",
                f"Set $f'(x) = 0$ and solve:",
                f"${latex(Eq(derivative, 0))}$",
                f"$x = {critical_pt}$"
            ),
        )
    
    elif problem_type == 2:
        # Type 2: Determine if a critical point is a max or min using second derivative (1100-1300)
        a_val = choice([-2, -1, 1, 2])
        h = nonzero(-5, 5)
        k = randint(-10, 10)
        expr = a_val * (x - h)**2 + k
        expr_expanded = expand(expr)
        
        second_deriv = diff(expr_expanded, x, 2)
        
        if a_val > 0:
            ans_text = "minimum"
        else:
            ans_text = "maximum"
        
        return problem(
            question=f"Does $f(x) = {latex(expr_expanded)}$ have a maximum or minimum at its critical point?",
            answer=ans_text,
            difficulty=(1100, 1300),
            topic="calculus/optimization",
            solution=steps(
                f"Find the second derivative: $f''(x) = {latex(second_deriv)}$",
                f"Since $f''(x) = {second_deriv}$ {'> 0' if a_val > 0 else '< 0'}$, the critical point is a {ans_text}"
            ),
        )
    
    elif problem_type == 3:
        # Type 3: Find maximum/minimum value of a simple quadratic (1100-1300)
        a_val = choice([-2, -1, 1, 2])
        h = nonzero(-5, 5)
        k = randint(-10, 10)
        expr = a_val * (x - h)**2 + k
        expr_expanded = expand(expr)
        
        if a_val > 0:
            question_text = "minimum"
        else:
            question_text = "maximum"
        
        ans = k
        
        return problem(
            question=f"Find the {question_text} value of $f(x) = {latex(expr_expanded)}$",
            answer=ans,
            difficulty=(1100, 1300),
            topic="calculus/optimization",
            solution=steps(
                f"Find the derivative: $f'(x) = {latex(diff(expr_expanded, x))}$",
                f"Set $f'(x) = 0$: $x = {h}$",
                f"Evaluate $f({h}) = {latex(expr_expanded.subs(x, h))} = {k}$",
                f"The {question_text} value is ${k}$"
            ),
        )
    
    else:
        # Type 4: Find where a linear function has its maximum on an interval (1200-1300)
        m = nonzero(-5, 5)
        b_val = randint(-10, 10)
        expr = m * x + b_val
        
        left = randint(0, 5)
        right = left + randint(3, 6)
        
        if m > 0:
            ans = right
            explanation = "increasing"
        else:
            ans = left
            explanation = "decreasing"
        
        return problem(
            question=f"At what value of $x$ in the interval $[{left}, {right}]$ does $f(x) = {latex(expr)}$ achieve its maximum?",
            answer=ans,
            difficulty=(1200, 1300),
            topic="calculus/optimization",
            solution=steps(
                f"Find the derivative: $f'(x) = {m}$",
                f"Since $f'(x) = {m}$ {'> 0' if m > 0 else '< 0'}$, the function is {explanation}",
                f"The maximum occurs at $x = {ans}$"
            ),
        )

emit(generate())

File: src/calculus_optimization_easy.jl
# calculus - optimization (easy)
# Generated: 2026-03-08T20:42:59.953007

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("calculus/optimization")
    
    prob_type = choice([1, 2, 3, 4, 5])
    
    if prob_type == 1
        # Maximum area of rectangle with fixed perimeter
        perimeter = randint(40, 120)
        half_p = perimeter // 2
        max_area = (half_p // 2)^2
        side_length = half_p // 2
        
        problem(
            question="A rectangle has a perimeter of $(perimeter) meters. What is the maximum possible area (in square meters)?",
            answer=max_area,
            difficulty=(700, 900),
            solution=steps(
                "Let width = \$w\$ and length = \$l\$. Perimeter: \$2w + 2l = $(perimeter)\$",
                "So \$l = $(half_p) - w\$. Area: \$A = w \\cdot l = w($(half_p) - w) = $(half_p)w - w^2\$",
                sol("Take derivative", diff(half_p*x - x^2, x)),
                "Set equal to zero: \$$(half_p) - 2w = 0\$, so \$w = $(side_length)\$ and \$l = $(side_length)\$",
                sol("Maximum area", max_area)
            ),
            time=120
        )
        
    elseif prob_type == 2
        # Find critical points of a quadratic
        a = nonzero(-5, -1)  # negative for maximum
        h = randint(-10, 10)
        k = randint(5, 25)
        
        expr = expand(a*(x - h)^2 + k)
        deriv = diff(expr, x)
        
        problem(
            question="Find the critical point of \$f(x) = $(tex(expr))\$",
            answer=h,
            difficulty=(700, 900),
            solution=steps(
                sol("Given", expr),
                sol("Find derivative", deriv),
                "Set \$f'(x) = 0\$: \$$(tex(deriv)) = 0\$",
                "Solve for \$x\$",
                sol("Critical point", h)
            ),
            time=90
        )
        
    elseif prob_type == 3
        # Maximum product of two numbers with fixed sum
        total = randint(30, 80)
        half = total // 2
        max_product = half^2
        
        problem(
            question="Two positive numbers have a sum of $(total). What is their maximum product?",
            answer=max_product,
            difficulty=(800, 1000),
            solution=steps(
                "Let the numbers be \$x\$ and \$$(total) - x\$",
                sol("Product", x*(total - x)),
                "Expand: \$P(x) = $(total)x - x^2\$",
                sol("Take derivative", diff(total*x - x^2, x)),
                "Set equal to zero: \$$(total) - 2x = 0\$, so \$x = $(half)\$",
                "Both numbers equal $(half)",
                sol("Maximum product", max_product)
            ),
            time=120
        )
        
    elseif prob_type == 4
        # Minimize sum of two positive numbers with fixed product
        product = choice([36, 48, 64, 81, 100])
        sqrt_prod = isqrt(product)
        min_sum = 2 * sqrt_prod
        
        problem(
            question="Two positive numbers have a product of $(product). What is the minimum value of their sum?",
            answer=min_sum,
            difficulty=(900, 1100),
            solution=steps(
                "Let the numbers be \$x\$ and \$\\frac{$(product)}{x}\$",
                sol("Sum", x + product/x),
                sol("Take derivative", diff(x + product/x, x)),
                "Set equal to zero: \$1 - \\frac{$(product)}{x^2} = 0\$",
                "Solve: \$x^2 = $(product)\$, so \$x = $(sqrt_prod)\$",
                "Both numbers equal $(sqrt_prod)",
                sol("Minimum sum", min_sum)
            ),
            time=150
        )
        
    else  # prob_type == 5
        # Maximum value of a cubic on an interval
        a = randint(-3, -1)
        b = randint(2, 8)
        c = randint(-15, 15)
        d = randint(-20, 20)
        
        expr = a*x^3 + b*x^2 + c*x + d
        deriv = diff(expr, x)
        
        # Find critical points
        coeffs_deriv = [3*a, 2*b, c]
        discriminant = (2*b)^2 - 4*(3*a)*c
        
        left_bound = randint(-5, 0)
        right_bound = randint(3, 8)
        
        # Evaluate at endpoints
        val_left = substitute(expr, x => left_bound)
        val_right = substitute(expr, x => right_bound)
        max_val = max(val_left, val_right)
        
        problem(
            question="Find the maximum value of \$f(x) = $(tex(expr))\$ on the interval \$[$(left_bound), $(right_bound)]\$",
            answer=max_val,
            difficulty=(1000, 1200),
            solution=steps(
                sol("Given", expr),
                sol("Find derivative", deriv),
                "Find critical points by setting \$f'(x) = 0\$",
                "Evaluate at endpoints: \$f($(left_bound)) = $(val_left)\$, \$f($(right_bound)) = $(val_right)\$",
                "Check critical points in interval (if any)",
                sol("Maximum value", max_val)
            ),
            time=180
        )
    end
end